homeassistant/components/thethingsnetwork/__init__.py

```diff
@@ -33,10 +27,65 @@
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Initialize of The Things Network component."""
-    conf = config[DOMAIN]
-    app_id = conf.get(CONF_APP_ID)
-    access_key = conf.get(CONF_ACCESS_KEY)
 
-    hass.data[DATA_TTN] = {TTN_ACCESS_KEY: access_key, TTN_APP_ID: app_id}
+    if DOMAIN in config:
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            "manual_migration",
+            breaks_in_ha_version="2021.10.0",
+            is_fixable=False,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="manual_migration",
+            translation_placeholders={"domain": DOMAIN},
+        )
 
     return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Establish connection with The Things Network."""
+
+    _LOGGER.debug(
+        "Set up %s at %s",
+        entry.data[CONF_APP_ID],
+        entry.data.get(CONF_HOSTNAME, TTN_API_HOSTNAME),
+    )
+
+    # Create coordinator to fetch TTN updates
+    entry.coordinator = coordinator = TTNCoordinator(hass, entry)
+
+    # Trigger the creation of entities for each supported platform
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Fetch all existing values in the TTN storage DB - NOTE: the free TTN only keeps 24 hours
+    await coordinator.async_config_entry_first_refresh()
```

Comment:
Please do this before setting up the platform or assigning it to hass.data

Reply:
I can do it but then I would need to add a call to each platform `async_setup_entry` to check for new instances instead of using the registered listener to process the initial data.
What do you think?